### ISSUE TITLE
amount of added noise for low number of microbatches is too harmful for performance.

### ISSUE THREAD

**Issue by ebagdasa** [tensorflow/privacy#46]
Sorry, it's probably a stupid question. If I understand correctly reducing number of microbatches from 256 to 32 will result in total norm of the sum to be less than 32*S, whereas for 256 batches it should be less than 256*S. Applying same amount of noise zS on a vector of a smaller norm results in a poorer performance. Can't we treat the gradient of each microbatch as a sum of gradients for k individual images and therefore clip it by kS, instead of mean and clip it by S?

**Comment by npapernot** [comment]
There is no stupid question. Microbatches were introduced to offer a trade-off between performance and per-example clipping. If we decrease the number of microbatches, we are no longer doing per-example clipping and so we are indeed losing some signal because multiple examples' gradients are combined before being clipped. Unfortunately, we can't multiply the clipping norm otherwise that would change the sensitivity of a training step to individual training examples (imagine for instance in the worst case that all gradients but one in the microbatch happened to be zero).

**Comment by ebagdasa** [comment]
oh, absolutely! Thanks a lot!

**Comment by ebagdasa** [comment]
How would it impact the guarantees, if the microbatches always contain the same set of images? Will we have per-microbatch level of privacy, similar to per-participant level in federated learning?
